interop-testing/src/main/java/io/grpc/testing/integration/XdsTestControlPlaneService.java

```diff
@@ -0,0 +1,354 @@
+      "type.googleapis.com/envoy.config.listener.v3.Listener";
+  private static final String ADS_TYPE_URL_RDS =
+      "type.googleapis.com/envoy.config.route.v3.RouteConfiguration";
+  private static final String ADS_TYPE_URL_CDS =
+      "type.googleapis.com/envoy.config.cluster.v3.Cluster";
+  private static final String ADS_TYPE_URL_EDS =
+      "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment";
+  private static final String HTTP_CONNECTION_MANAGER_TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3"
+          + ".HttpConnectionManager";
+
+  private static final String rdsName = "route-config.googleapis.com";
+  private static final String clusterName = "cluster0";
+  private static final String edsName = "eds-service-0";
+  private final ImmutableMap<String, Listener> ldsResources;
+  private final ImmutableMap<String, RouteConfiguration> rdsResources;
+  private final ImmutableMap<String, Cluster> cdsResources;
+  private final ImmutableMap<String, ClusterLoadAssignment> edsResources;
+  private int versionNo = 0;
+
+  /**
+   * Create a control plane service for testing, with static xds configurations.
+   */
+  public XdsTestControlPlaneService(XdsTestControlPlaneConfig config) {
+    this.ldsResources = ImmutableMap.of(config.apiListener.getName(), config.apiListener,
+        config.tcpListener.getName(), config.tcpListener);
+    this.rdsResources = ImmutableMap.of(rdsName, config.rds);
+    this.cdsResources = ImmutableMap.of(clusterName, config.cds);
+    this.edsResources = ImmutableMap.of(edsName, config.eds);
+  }
+
+  public static class XdsTestControlPlaneConfig {
+    Listener tcpListener;
+    Listener apiListener;
+    RouteConfiguration rds;
+    Cluster cds;
+    ClusterLoadAssignment eds;
+
+    /**
+     * Provide control plane xds configurations.
+     */
+    public XdsTestControlPlaneConfig(Listener tcpListener, Listener apiListener,
+         RouteConfiguration rds, Cluster cds, ClusterLoadAssignment eds) {
+      this.tcpListener = tcpListener;
+      this.apiListener = apiListener;
+      this.rds = rds;
+      this.cds = cds;
+      this.eds = eds;
+    }
+  }
+
+  @Override
+  public StreamObserver<DiscoveryRequest> streamAggregatedResources(
+      final StreamObserver<DiscoveryResponse> responseObserver) {
+    final StreamObserver<DiscoveryRequest> requestObserver =
+        new StreamObserver<DiscoveryRequest>() {
+      @Override
+      public void onNext(final DiscoveryRequest value) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            logger.log(Level.FINEST, "control plane received request {0}", value);
+            if (!value.getResponseNonce().isEmpty()) {
+              logger.log(Level.FINEST, "control plane received ack for resource:  {0}",
+                  value.getResourceNamesList());
+              return;
+            }
+            if (value.hasErrorDetail()) {
```

Comment:
For request that is ACKing previous response, i.e., value.versionInfo == current versionNo , seems it shouldn't send a response immediately, otherwise there will be another ACK immediately. Either stop responding or delay responding the next version.

Reply:
I see you are checking nonce to determine if it is an ACK.